### PR TITLE
Fix crash of COPY FROM on AO/AOCO/PAX partition tables.

### LINKED ISSUE #547: [Bug] COPY FROM crash on partition tables with partitions > 32 and data of special order
### Cloudberry Database version

```sql
                                                                                                          version

-----------------------------------------------------------------------------------------------------------------------
-----------------------------------------------------------------------------------------------------
 PostgreSQL 14.4 (Cloudberry Database 1.5.4+dev.106.g057f9d27525 build dev) on x86_64-pc-linux-gnu, compiled by gcc (Ub
untu 11.3.0-1ubuntu1~22.04.1) 11.3.0, 64-bit compiled on Jul 31 2024 17:59:22 (with assert checking)
(1 row)
```

### What happened

COPY FROM coredump as:
img_v3_02da_4f85f0ef-4f4f-43a5-8cf5-40ad52a6883g

### What you think should happen instead

_No response_

### How to reproduce

1. prepare a data file:
```shell
for i in {1..967}; do echo 2, 1 >> /tmp/a.csv;done
for i in {1..33}; do echo 2, $i >> /tmp/a.csv;done
echo 2, 1 >> /tmp/a.csv
```
2. create a partitioned table with more than 32 partitions:
```sql
gpadmin=# create table px(a int, b int) partition by range(b) (start(1) end(34) every(1)) using ao_column distributed by (a);
CREATE TABLE
```
3. COPY FROM the data file
```sql
gpadmin=# copy px from '/tmp/a.csv' (format csv);
ERROR:  MPP detected 1 segment failures, system is reconnected
```
then will see a coredump file on segment 0

### Operating System

Ubuntu

### Anything else

_No response_

### Are you willing to submit PR?

- [X] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://cloudberrydb.org/community/coc).

### PR DESCRIPTION
**Pull request by avamingli**
Fix #547 

AO/AOCO tables clean up state in am routine finish_bulk_insert() which may be called multiple times during COPY FROM

For a partitioned table has more than 32 partitions(MAX_PARTITION_BUFFERS), when copy partition buffer is flushed(enough tuples or bytes in buffer), some partitions would be cleaned up with table_finish_bulk_insert().
And if the next input tuple belongs to that partition we just flushed out on the same segment, crash will happen.

copy_from drawio

COPY partitioned table FROM xxx..

T0:   QD route data to segments and partitions, buffer with 32 partitions on segment0 is not full.
T1:    buffer on segment0 with 33 partition is full, it's flushed and swap out p1 table(finish_bulk_insert called, p1 insert state is released).
T2:   QD route new data belongs to partition1(p1) on segment0, then crash happened.

Authored-by: Zhang Mingli avamingli@gmail.com

fix #ISSUE_Number

---

### Change logs

_Describe your change clearly, including what problem is being solved or what feature is being added._

_If it has some breaking backward or forward compatibility, please clary._

### Why are the changes needed?

_Describe why the changes are necessary._

### Does this PR introduce any user-facing change?

_If yes, please clarify the previous behavior and the change this PR proposes._

### How was this patch tested?

_Please detail how the changes were tested, including manual tests and any relevant unit or integration tests._

### Contributor's Checklist

Here are some reminders and checklists before/when submitting your pull request, please check them:

- [ ] Make sure your Pull Request has a clear title and commit message. You can take [git-commit](https://github.com/cloudberrydb/cloudberrydb/blob/main/.gitmessage) template as a reference.
- [ ] Sign the Contributor License Agreement as prompted for your first-time contribution(*One-time setup*).
- [ ] Learn the [coding contribution guide](https://cloudberrydb.org/contribute/code), including our code conventions, workflow and more.
- [ ] List your communication in the [GitHub Issues](https://github.com/cloudberrydb/cloudberrydb/issues) or [Discussions](https://github.com/orgs/cloudberrydb/discussions) (if has or needed).
- [ ] Document changes.
- [ ] Add tests for the change
- [ ] Pass `make installcheck`
- [ ] Pass `make -C src/test installcheck-cbdb-parallel`
- [ ] Feel free to request `cloudberrydb/dev` team for review and approval when your PR is ready🥳

### CODE REVIEW

**Review by my-ship-it — APPROVED** [review]

File: src/backend/access/aocs/aocsam_handler.c [hunk 1]
```diff
@@ -1174,7 +1174,14 @@ aoco_tuple_lock(Relation relation, ItemPointer tid, Snapshot snapshot,
...
 aoco_finish_bulk_insert(Relation relation, int options)
 {
-	aoco_dml_finish(relation, CMD_INSERT);
+	/*
+	 * https://github.com/cloudberrydb/cloudberrydb/issues/547
+	 * Do not clean up context or resource here(ex: aoco_dml_finish),
+	 * finish_bulk_insert routine will be called more than once during
+	 * COPY FROM if the partition buffer is flushed but COPY is not finished.
+	 *
+	 * FIXME: Do we need to keep this am routine? Seems no used anymore.
```
Comment (gfphoenix78) [inline src/backend/access/aocs/aocsam_handler.c:1183]: https://github.com/cloudberrydb/cloudberrydb/blob/main/src/backend/commands/copyfrom.c#L2534
Remove the call `table_finish_bulk_insert`?

Comment (avamingli) [reply src/backend/access/aocs/aocsam_handler.c:1183]: +1, that call is useless as we have already call related functions like aoco_dml_finish() 
https://github.com/cloudberrydb/cloudberrydb/blob/main/src/backend/commands/copyfrom.c#L2433
I think we are ok to  remove it.

Comment (avamingli) [reply src/backend/access/aocs/aocsam_handler.c]: Done.

**Comment by avamingli** [comment]
table_finish_bulk_insert is still needed in many places to clean up resource, GPDB use aoco_dml_init() to init resource and table_finish_bulk_insert() to clean up, ex: create as.

Move fix to copyFrom to work around.

**Comment by avamingli** [comment]
Make a new approach to fix, @my-ship-it @gfphoenix78 , PTAL.

File: src/backend/commands/copyfrom.c [hunk 2]
```diff
@@ -565,6 +565,19 @@ CopyMultiInsertBufferCleanup(CopyMultiInsertInfo *miinfo,
 	for (i = 0; i < MAX_BUFFERED_TUPLES && buffer->slots[i] != NULL; i++)
 		ExecDropSingleTupleTableSlot(buffer->slots[i]);
 
+	if (AMHandlerIsAO(buffer->resultRelInfo->ri_RelationDesc->rd_amhandler))
```
Comment (my-ship-it) [inline src/backend/commands/copyfrom.c:568]: Is it possible PAX and storage am still fail?

Comment (avamingli) [reply src/backend/commands/copyfrom.c:568]: Will have a test later

Comment (avamingli) [reply src/backend/commands/copyfrom.c:568]: @jiaqizho has a test, confirm pax  has this issue too.

img_v3_02di_1ee93136-f9da-4daf-bab0-5b2427e9518g

Comment (avamingli) [reply src/backend/commands/copyfrom.c]: Fixed.

**Review by my-ship-it — APPROVED** [review]

**Pull request merged**